### PR TITLE
Added a new best practices for custom form field types

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | #5077 |

### CODE REVIEW

File: best_practices/forms.rst [hunk 1]
```diff
@@ -207,3 +207,21 @@ Second, we recommend using ``$form->isSubmitted()`` in the ``if`` statement
...
 like the form is *always* processed (even on the GET request).
+
+Custom Form Field Types
+-----------------------
+
+.. best-practice::
+
+    Add the ``app.`` prefix to your custom form field types to avoid collisions.
+
+Custom form field types extend from the ``AbstractType`` class, which defines the
```
Comment (xabbuh) [inline best_practices/forms.rst]: "extend the" or "inherit from" I guess.

File: best_practices/forms.rst [hunk 2]
```diff
@@ -207,3 +207,21 @@ Second, we recommend using ``$form->isSubmitted()`` in the ``if`` statement
...
+``getName()`` method to configure the name of that form type. These names must
+be unique in the application.
+
+If a custom form type uses the same name as any of the Symfony's built-in form
+types, it will override it. The same happens when the custom form type matches
+any of the types defined by the third-party bundles installed in you application.
```
Comment (xabbuh) [inline best_practices/forms.rst]: your

File: best_practices/forms.rst [hunk 3]
```diff
@@ -207,3 +207,21 @@ Second, we recommend using ``$form->isSubmitted()`` in the ``if`` statement
 for clarity. This isn't technically needed, since ``isValid()`` first calls
 ``isSubmitted()``. But without this, the flow doesn't read well as it *looks*
 like the form is *always* processed (even on the GET request).
+
+Custom Form Field Types
+-----------------------
+
+.. best-practice::
+
+    Add the ``app.`` prefix to your custom form field types to avoid collisions.
```
Comment (stof) [inline best_practices/forms.rst]: dots are forbidden in type names, so your suggestion is really bad. Recommend ``app_`` instead.

for reference, type names are validated as ``/^[a-z0-9_]*$/i``

**Comment by WouterJ** [comment]
I see a little problem here: We already have quite a few best practices just recommending an `app` prefix (some with a dot, some with an underscore, some with a namespace, etc.). Going like that, don't we end up writing a best practice about such a prefix for all customizable things in Symfony? (e.g. use the `app:` prefix for commands, ...).

I think it is better if we just replace all these best practice with one best practice: Always prefix with `app`. We add this to the AppBundle article. If we think the suffix of this might be confusing (dot, underscore, dash, colon, ...) we can add a list of examples.

**Comment by weaverryan** [comment]
@WouterJ I hear what you're saying, it might not be a bad idea. But what *are* the places we prefix? I can think of:

1) form type names
2) service ids

I'm probably missing a few, but I'm not sure.

**Comment by weaverryan** [comment]
I've merged this in - there was no disagreement here or on the issue, and I can't think of why there would be any. But @WouterJ, I don't want to lose your point -  so please reply and we can definitely reformat how we're communicating the best practices. Javier, I updated #5077's description to have 2 checkboxes - the second one is to update the documentation to add the app_ prefix. Thanks!

**Pull request merged**